File: j-LIB-core/src/main/java/io/github/dot166/jlib/rss/RSSFeedAggregator.java
package io.github.dot166.jlib.rss;

import static io.github.dot166.jlib.utils.DateUtils.convertDateToEpochSeconds;
import static io.github.dot166.jlib.utils.DateUtils.convertFromCommonFormats;

import com.prof18.rssparser.model.RssChannel;
import com.prof18.rssparser.model.RssItem;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Objects;

class RSSFeedAggregator {

    static RssChannel buildAllFeeds(String[] rssUrls, RSSViewModel viewModel) {
        List<RssItem> items = new ArrayList<>();

        for (String url : rssUrls) {
            RssChannel channel = viewModel.fetchFeedWithoutViewModel(url);

            if (!channel.getItems().isEmpty()
                    && Objects.equals(channel.getItems().get(0).getTitle(), "Error Handler")
                    && Objects.equals(channel.getItems().get(0).getSourceName(), "jLib")) {
                break;
            }

            items.addAll(channel.getItems());
        }

        items.sort(Comparator.comparingLong(
                item -> convertDateToEpochSeconds(convertFromCommonFormats(((RssItem)item).getPubDate()))
        ).reversed());

        return new RssChannel("All Feeds", null, null, null, null, null, items, null, null);
    }
}

File: j-LIB-core/src/main/java/io/github/dot166/jlib/rss/RSSFeedAggregator.kt
package io.github.dot166.jlib.rss

import com.prof18.rssparser.model.RssChannel
import com.prof18.rssparser.model.RssItem
import io.github.dot166.jlib.utils.DateUtils.convertDateToEpochSeconds
import io.github.dot166.jlib.utils.DateUtils.convertFromCommonFormats

internal object RSSFeedAggregator {
    fun buildAllFeeds(rssUrls: Array<String?>, viewModel: RSSViewModel): RssChannel {
        val items: MutableList<RssItem> = ArrayList()

        for (url in rssUrls) {
            if (url != null) {
                val channel = viewModel.fetchFeedWithoutViewModel(url)

                if (!channel.items.isEmpty() && channel.items[0].title == "Error Handler"
                    && channel.items[0].sourceName == "jLib"
                ) {
                    break
                }

                items.addAll(channel.items)
            }
        }

        items.sortWith(
            Comparator.comparingLong<Any?> { item: Any? ->
                convertDateToEpochSeconds(
                    convertFromCommonFormats((item as RssItem).pubDate)
                )
            }.reversed())

        return RssChannel("All Feeds", null, null, null, null, null,
            items as List<RssItem>, null, null)
    }
}
